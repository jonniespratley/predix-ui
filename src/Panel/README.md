The `<Panel/>` component provides a container for placing controls or content. Can be fixed or relatively positioned, and persistent or shown and hidden.



## Usage

```react
state:
  opened: false
---
const demoStyles = {
	position: 'relative',
	overflow: 'hidden',
	minHeight: '400px',
	width: '100%',
	boxSizing: 'border-box',
	border: '1px solid gray',
	padding: '1rem'
};

const { 
  Button,
  Panel 
} = PxReact;

<div style={demoStyles}>
    <p>Use the toggles in properties to invoke the panel.</p>
    <Button onClick={() => setState({opened: !state.opened})}>
      {state.opened ? 'Close' : 'Open'} 
    </Button>
    
    <Panel 
      position='right'
      background='light'
      opened={state.opened}>
      <p>This can be any type of content.</p>
    </Panel>
  </div>
```


## Properties

```table
span: 6
rows:
  
  - Name: background
    Type: String
    Description: Whether to display the panel with a light, medium, or dark background. These terms are relative, and can be used in conjunction with px-theme, px-dark-theme, or your custom theme or CSS variables to toggle between 3 different background colors.
  
  - Name: fixed
    Type: bool
    Description: If set to true, the panel will have position:fixed so it will be attached to the browser window instead of its parent container.
  
  - Name: floating
    Type: bool
    Description: If set to true, the panel will appear with an offset relative to the screen / container edges. Can be overridden or customized using the CSS variables.
  
  - Name: minimizable
    Type: bool
    Description: If set to true, the panel will not fully collapse when closed. It will appear in a minimized state. You can use the content slot minimized to determine what will appear inside the panel while minimized (likely an icon or button for fully opening the panel).


  - Name: opened
    Type: bool
    Description: Whether the panel is currently open (expanded). 
  
  - Name: persistent
    Type: bool
    Description: If set to true, the panel will be opened and calls to the close() method will be ignored. Mutating the opened property will still force a close of the panel.
  
  - Name: position
    Type: String
    Description: Where to place the panel - one of `top`, `bottom`, `left`, or `right`.
```



### Styling
The following custom properties are available for styling:

```table
span: 6
rows:
  - Custom Property: --px-panel-size
    Description: Width (left/right) or height (top/bottom) of the panel
  - Custom Property: --px-panel-size--minimized
    Description: Width (left/right) or height (top/bottom) of the panel when "minimized"
  - Custom Property: --px-panel-bg-color--light
    Description: Background color for "light" panels
  - Custom Property: --px-panel-bg-color--medium
    Description: Background color for "medium" panels
  - Custom Property: --px-panel-bg-color--dark
    Description: Background color for "dark" panels
  - Custom Property: --px-panel-offset--top
    Description: Offset for top of panel when "floating"
  - Custom Property: --px-panel-offset--right
    Description: Offset for right side of panel when "floating"
  - Custom Property: --px-panel-offset--bottom
    Description: Offset for bottom of panel when "floating"
  - Custom Property: --px-panel-offset--left
    Description: Offset for left side of panel when "floating"
  - Custom Property: --px-panel-z-index
    Description: Z-index of the panel - should be adjusted if multiple overlapping panels are possible
```