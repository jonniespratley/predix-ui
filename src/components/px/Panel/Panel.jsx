import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const Panel = styled.div`
  overflow:hidden;
  display: block;
  position: absolute;
  z-index: var(--px-panel-z-index, 100);
  box-shadow: var(--px-panel-box-shadow, -1px -1px 3px rgba(0,0,0,0.2), 1px  1px 3px rgba(0,0,0,0.2));
  background: var(--px-panel-background-color, white);
  border: var(--px-panel-border, none);
  transition: var(--px-panel-transition, none);

  ${props => props.background === 'light' && css`
    background: var(--px-panel-bg-color--light, #fff);
  `}

  ${props => props.background === 'medium' && css`
    background: var(--px-panel-bg-color--medium, #eee);
  `}

  ${props => props.background === 'dark' && css`
    background: var(--px-panel-bg-color--dark, #ddd);
  `}

  ${props => props.light && css`
    position: fixed;
  `}

  ${props => props.fixed && css`
    position: fixed;
  `}

  ${props => props.minimizable && css`
    width: var(--px-panel-size--minimized,4rem);
  `}

  ${props => props.persistent && css`
    position: relative;
    flex: 0 1 320px;
  `}

  ${props => (props.position === 'left' || props.position === 'right') && css`
    height: auto;
    width: 0;
    top: 0;
    bottom: 0;
    transition: var(--px-panel-transition, width 0.4s cubic-bezier(.78,.13,.16,.87));
    white-space: nowrap;

    ${props.opened && css`
      width: var(--px-panel-size, 320px);
      overflow-y: auto;
      ${props.fullSize && css`
        width: 100%;
        overflow-y: auto;
      `}
    `}
  `}

   ${props => props.position === 'left' && css`
    left: 0;
    border-right: 1px solid var(--px-panel-border-color,gray);

    ${props.floating && css`
      border-right: none;
      left: var(--px-panel-offset--left, 2rem);
      top: var(--px-panel-offset--top, 2rem);
      bottom: var(--px-panel-offset--bottom, 2rem);
    `}
  `}

  ${props => props.position === 'right' && css`
    right: 0;
    border-left: 1px solid var(--px-panel-border-color,gray);

    ${props.floating && css`
      border-left: none;
      right: var(--px-panel-offset--right, 2rem);
      top: var(--px-panel-offset--top, 2rem);
      bottom: var(--px-panel-offset--bottom, 2rem);
    `}
  `}

  ${props => (props.position === 'top' || props.position === 'bottom') && css`
    width: auto;
    height: 0;
    right: 0;
    left: 0;
    transition: var(--px-panel-transition, height 0.4s cubic-bezier(.78,.13,.16,.87));

    ${props.opened && css`
      height: var(--px-panel-size, 320px);
      overflow-y: auto;
    `}
  `}

  ${props => props.position === 'top' && css`
    top: 0;
    border-bottom: 1px solid var(--px-panel-border-color, gray);

    ${props.fullSize && css`
      right: var(--px-panel-offset--right, 2rem);
      left: var(--px-panel-offset--left, 2rem);
      top: var(--px-panel-offset--bottom, 2rem);
    `}
  `}

  ${props => props.position === 'bottom' && css`
    bottom: 0;
    border-top: 1px solid var(--px-panel-border-color, gray);

    ${props.fullSize && css`
      right: var(--px-panel-offset--right, 2rem);
      left: var(--px-panel-offset--left, 2rem);
      bottom: var(--px-panel-offset--bottom, 2rem);
    `}
  `}
`;

Panel.displayName = 'Panel';
Panel.defaultProps = {
  className: 'px-panel'
};

const PanelContent = styled.div`
  display: block;
  ${props => props.opened && css`
    display: block;
    height: 100%;
  `}
`;
PanelContent.displayName = 'PanelContent';
PanelContent.defaultProps = {
  className: 'px-panel-content'
};
/**
 * Panel component
 */
const Component = ({
  style,
  className,
  background,
  position,
  fullSize,
  opened,
  fixed,
  persistent,
  floating,
  minimizable,
  children
}) => (
  <Panel
    persistent={persistent}
    background={background}
    style={style}
    className={className}
    position={position}
    minimizable={minimizable}
    fullSize={fullSize}
    floating={floating}
    opened={opened}
    fixed={fixed}
  >
    <PanelContent>{children}</PanelContent>
  </Panel>
);

Component.defaultProps = {
  style: null,
  className: null,
  background: 'light',
  position: 'right',
  fullSize: false,
  opened: false,
  fixed: false,
  persistent: false,
  floating: false,
  minimizable: false,
  children: null
};

Component.propTypes = {
  style: PropTypes.objectOf(PropTypes.string),
  className: PropTypes.string,
  background: PropTypes.string,
  position: PropTypes.string,
  fullSize: PropTypes.bool,
  opened: PropTypes.bool,
  fixed: PropTypes.bool,
  persistent: PropTypes.bool,
  floating: PropTypes.bool,
  minimizable: PropTypes.bool,
  children: PropTypes.node
};

export default Component;
