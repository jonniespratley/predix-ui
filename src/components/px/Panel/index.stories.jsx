import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, select } from '@storybook/addon-knobs';

import Panel from './';
import Card from '../Card';

const demoStyles = {
  position: 'relative',
  overflow: 'hidden',
  minHeight: '400px',
  width: '100%',
  boxSizing: 'border-box',
  border: '1px solid gray',
  padding: '1rem'
};

const positions = [
  'top',
  'right',
  'left',
  'bottom'
];

const bgColors = [
  'light',
  'medium',
  'dark'
];

const actions = () => (<div>Expand</div>);

storiesOf('Components / Panel', module)
  .addDecorator(withKnobs)
  .add('default', () => (
    <div style={demoStyles}>
      <p>Use the toggles in properties to invoke the panel.</p>
      <Panel
        position={select('position', positions, 'right')}
        background={select('background', bgColors, 'light')}
        fullSize={boolean('fullSize', false)}
        minimizable={boolean('minimizable', false)}
        floating={boolean('floating', false)}
        fixed={boolean('fixed', false)}
        persistent={boolean('persistent', false)}
        opened={boolean('opened', true)}
      >

        <p>This can be any type of content.</p>
        <p>A form or a map</p>

      </Panel>
    </div>

  ))
  .add('with multiple', () => (
    <div style={demoStyles}>
      <p>Here is multiple panels</p>
      <Panel
        id="p1"
        opened={boolean('p1Opened', true)}
        fullSize={boolean('p1FullSize', false)}
        persistent
      >
        <Card headerText="Main Card">This is p1 content</Card>
      </Panel>
      <Panel
        id="p2"
        persistent
        fullSize={boolean('p2FullSize', false)}
        opened={boolean('p2Opened', true)}
        background="medium"
      >
        <Card headerText="Secondary Card">This is p2 content</Card>
      </Panel>
      <Panel
        id="p3"
        persistent
        fullSize={boolean('p3FullSize', false)}
        opened={boolean('p3Opened', false)}
        background="dark"
      >
        <Card headerText="Action Card" actions={actions}>This is p3 content</Card>
      </Panel>
    </div>
  ));

