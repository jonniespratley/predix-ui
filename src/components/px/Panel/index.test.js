import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import Panel from './';

describe('Panel', () => {
  test('should render', () =>{
    const wrapper = shallow(
      <Panel position='left'>
        Panel content goes here.
      </Panel>
    );
    expect(wrapper.props().position).toEqual('left');
  });

  test('renders block when visible', () => {
    const tree = renderer.create(<Panel position='left' visible />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('display', 'block');
  });
  test('fixed renders correctly', () => {
    const tree = renderer.create(<Panel position='right' fixed visible />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('position', 'fixed');
  });
});
