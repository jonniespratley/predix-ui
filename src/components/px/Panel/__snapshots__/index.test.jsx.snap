// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel fixed renders correctly 1`] = `
.c0 {
  overflow: hidden;
  display: block;
  position: absolute;
  z-index: var(--px-panel-z-index,100);
  box-shadow: var(--px-panel-box-shadow,-1px -1px 3px rgba(0,0,0,0.2),1px 1px 3px rgba(0,0,0,0.2));
  background: var(--px-panel-background-color,white);
  border: var(--px-panel-border,none);
  -webkit-transition: var(--px-panel-transition,none);
  transition: var(--px-panel-transition,none);
  background: var(--px-panel-bg-color--light,#fff);
  position: fixed;
  height: auto;
  width: 0;
  top: 0;
  bottom: 0;
  -webkit-transition: var(--px-panel-transition,width 0.4s cubic-bezier(.78,.13,.16,.87));
  transition: var(--px-panel-transition,width 0.4s cubic-bezier(.78,.13,.16,.87));
  white-space: nowrap;
  right: 0;
  border-left: 1px solid var(--px-panel-border-color,gray);
}

.c1 {
  display: block;
}

<div
  className="c0"
  style={null}
>
  <div
    className="px-panel-content c1"
  />
</div>
`;

exports[`Panel renders block when visible 1`] = `
.c0 {
  overflow: hidden;
  display: block;
  position: absolute;
  z-index: var(--px-panel-z-index,100);
  box-shadow: var(--px-panel-box-shadow,-1px -1px 3px rgba(0,0,0,0.2),1px 1px 3px rgba(0,0,0,0.2));
  background: var(--px-panel-background-color,white);
  border: var(--px-panel-border,none);
  -webkit-transition: var(--px-panel-transition,none);
  transition: var(--px-panel-transition,none);
  background: var(--px-panel-bg-color--light,#fff);
  height: auto;
  width: 0;
  top: 0;
  bottom: 0;
  -webkit-transition: var(--px-panel-transition,width 0.4s cubic-bezier(.78,.13,.16,.87));
  transition: var(--px-panel-transition,width 0.4s cubic-bezier(.78,.13,.16,.87));
  white-space: nowrap;
  left: 0;
  border-right: 1px solid var(--px-panel-border-color,gray);
}

.c1 {
  display: block;
}

<div
  className="c0"
  style={null}
>
  <div
    className="px-panel-content c1"
  />
</div>
`;
